feat(cli): Add research communication opt-in capability

yashv6655 opened the pull request:
## Moved to https://github.com/google-gemini/gemini-cli/pull/6616

Resolves https://github.com/google-gemini/gemini-cli/issues/4812

yashv6655 commented:
/gemini review

richieforeman commented on packages/cli/src/ui/commands/feedbackCommand.test.ts:
```diff
@@ -0,0 +1,160 @@
+  });
+
+  it('should accept feedback and log telemetry event when opted in with args', async () => {
+    const mockConfig: Partial<Config> = { getTelemetryEnabled: () => true };
```
Can you prefer the `makeMockConfig` helper

ref: https://github.com/google-gemini/gemini-cli/blob/main/packages/core/src/test-utils/config.ts

richieforeman commented on packages/cli/src/ui/commands/feedbackCommand.test.ts:
```diff
@@ -0,0 +1,160 @@
+
+    // Verify telemetry logging was called
+    expect(vi.mocked(logResearchFeedback)).toHaveBeenCalledOnce();
+    const [configArg, eventArg] = vi.mocked(logResearchFeedback).mock.calls[0];
```
nit: Please prefer `expect(logResearchFeedback).toHaveBeenCalledWith(mockConfig, researchFeedbackEvent)`

richieforeman commented on packages/cli/src/ui/commands/feedbackCommand.test.ts:
```diff
@@ -0,0 +1,160 @@
+  });
+
+  it('should always log research feedback when config exists', async () => {
+    const mockConfig: Partial<Config> = { getTelemetryEnabled: () => false };
```
Same commentary here above (makeMockConfig)

richieforeman commented on packages/core/src/telemetry/clearcut-logger/event-metadata-key.ts:
```diff
@@ -234,4 +234,26 @@ export enum EventMetadataKey {
+  // ===========================================================================
+
+  // Logs the research opt-in status (true/false)
+  GEMINI_CLI_RESEARCH_OPT_IN_STATUS = 64,
```
As you've discovered, this file is a bit of a pain in that it needs to be kept in sync vs an internal system.  You'll need a googler to send that change internally for you.  I'm happy to do this once we get thru another review cycle. 

FYI: We might have to change these numbers, but this looks fine for the moment.

richieforeman commented on packages/core/src/telemetry/constants.ts:
```diff
@@ -24,3 +26,19 @@ export const METRIC_API_REQUEST_LATENCY = 'gemini_cli.api.request.latency';
+  if (!content) {
+    return undefined;
+  }
+  return Array.from(content).slice(0, FEEDBACK_CONTENT_MAX_LENGTH).join('');
```
This looks to me like a fancy String.substr.  Can you just prefer a substr?

Something like:

`(content ?? '').substr(0, MAX_LENGTH);`

richieforeman commented on packages/core/src/telemetry/loggers.test.ts:
```diff
@@ -804,4 +807,46 @@ describe('loggers', () => {
+      // Clear any previous mock calls
+      mockLogger.emit.mockClear();
+      
+      const mockConfig = {
```
As above, prefer the `makeMockConfig` helper plz :)

richieforeman commented on packages/core/src/telemetry/types.ts:
```diff
@@ -379,6 +379,49 @@ export class KittySequenceOverflowEvent {
   }
 }
 
+export class ResearchOptInEvent implements BaseTelemetryEvent {
```
I know there's a bit of inconsistency in this file (I'm sorry, I'll be fixing it and making it consistent over the next couple of days).  I think you'll find this usage to be a bit easier to work with.

However, we now prefer factory functions + pojos over these classes.  See here for an example:
https://github.com/google-gemini/gemini-cli/blob/main/packages/core/src/telemetry/types.ts#L299-L319

yashv6655 commented:
@richieforeman I have addressed your concerns and pushed them. I had to open a new PR because I was getting too many merge conflicts errors from main and it was causing a crazy amount of compilation errors. I moved it to here with all your suggestions: https://github.com/google-gemini/gemini-cli/pull/6616
